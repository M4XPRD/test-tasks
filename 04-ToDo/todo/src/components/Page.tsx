import React from 'react';
import useTodo from '../hooks/todoHook';
import InputForm from './InputForm';
import TasksList from './TasksList';
import FilterBar from './FilterBar';

const Page = () => {
  const todo = useTodo();
  const { dropdownActive } = todo;

  return (
    <div className="page">
      <h1 className="page__h1">todos</h1>
      <main className="page__form">
        <InputForm />
        {dropdownActive && (<FilterBar type="main" />)}
        <TasksList />
      </main>
      <FilterBar type="footer" />
    </div>
  );
};

export default Page;
