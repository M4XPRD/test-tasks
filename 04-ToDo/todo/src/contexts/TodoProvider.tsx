import {
  useState, useMemo, ReactNode, ChangeEvent, FormEvent,
} from 'react';
import _ from 'lodash';
import TodoContext from './TodoContext';

interface TodoProviderProps {
  children: ReactNode;
}

const TodoProvider = ({ children }: TodoProviderProps) => {
  const [todos, setTodos] = useState<Array<{ text: string; completed: boolean; id: number }>>([]);
  const [inputValue, setInputValue] = useState<string>('');
  const [activeFilter, setActiveFilter] = useState<string>('all');
  const [dropdownActive, setDropdownActive] = useState<boolean>(false);

  const handleDropdown = () => {
    setDropdownActive(!dropdownActive);
  };

  const handleInputChange = (e: ChangeEvent<HTMLInputElement>) => {
    setInputValue(e.target.value);
  };

  const handleFormSubmit = (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    if (inputValue.trim() !== '') {
      const newTodo = {
        text: inputValue.trim(),
        completed: false,
        id: Number(_.uniqueId()),
      };
      setTodos([...todos, newTodo]);
      setInputValue('');
    }
  };

  const handleTodoClick = (index: number) => {
    const updatedTodos = [...todos];
    updatedTodos[index].completed = !updatedTodos[index].completed;
    setTodos(updatedTodos);
  };

  const handleDeleteClick = (index: number) => {
    const updatedTodos = todos.filter((todo, todoIndex) => todoIndex !== index);
    setTodos(updatedTodos);
  };

  const renderTodos = () => {
    switch (activeFilter) {
      case 'all':
        return todos.length;
      case 'active': {
        const activeTodos = todos.filter((todo) => !todo.completed);
        return activeTodos.length;
      }
      case 'completed': {
        const completedTodos = todos.filter((todo) => todo.completed);
        return completedTodos.length;
      }
      default:
        return 0;
    }
  };

  const handleTodosFilter = (type: string) => {
    switch (type) {
      case 'all':
        setActiveFilter('all');
        break;
      case 'active':
        setActiveFilter('active');
        break;
      case 'completed':
        setActiveFilter('completed');
        break;
      case 'clear': {
        setActiveFilter('clear');
        const unfinishedTodos = todos.filter(
          ({ completed }) => completed === false,
        );
        setTodos(unfinishedTodos);
      }
        setActiveFilter('all');
        break;
      default:
        break;
    }
  };

  const providedData = useMemo(() => ({
    todos,
    setTodos,
    inputValue,
    setInputValue,
    activeFilter,
    setActiveFilter,
    dropdownActive,
    setDropdownActive,
    handleTodosFilter,
    handleDropdown,
    handleInputChange,
    handleFormSubmit,
    handleTodoClick,
    handleDeleteClick,
    renderTodos,
  }), [
    todos,
    setTodos,
    inputValue,
    setInputValue,
    activeFilter,
    setActiveFilter,
    dropdownActive,
    setDropdownActive,
    handleTodosFilter,
    handleDropdown,
    handleInputChange,
    handleFormSubmit,
    handleTodoClick,
    handleDeleteClick,
    renderTodos,
  ]);

  return (
    <TodoContext.Provider value={providedData}>
      {children}
    </TodoContext.Provider>
  );
};

export default TodoProvider;
