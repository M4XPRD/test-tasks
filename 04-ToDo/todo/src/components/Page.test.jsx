import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { act } from 'react-dom/test-utils';
import Page from './Page';
import TodoProvider from '../contexts/TodoProvider';

test('Page components renders', () => {
  render(<Page />);
  const findElement = screen.getByText(/todos/i);
  expect(findElement).toBeInTheDocument();
});

test('Page snapshot test', () => {
  const element = render(<Page />);
  expect(element).toMatchSnapshot();
});

test('Tasks are shown by default on Page', () => {
  render(<Page />);
  const tasks = screen.getByTestId('tasksList');
  expect(tasks).not.toHaveClass('page__form-ul_hide');
});

test('Tasks are hidden on Page when dropdown button is clicked', async () => {
  render(
    <TodoProvider>
      <Page />
    </TodoProvider>,
  );
  const dropdownButton = screen.getByTestId('dropdownButton');
  act(() => {
    userEvent.click(dropdownButton);
  });
  const tasks = screen.getByTestId('tasksList');
  expect(tasks).toHaveClass('page__form-ul_hide');
});
