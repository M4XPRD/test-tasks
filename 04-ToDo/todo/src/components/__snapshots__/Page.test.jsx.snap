// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page snapshot test 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="page"
      >
        <h1
          class="page__h1"
        >
          todos
        </h1>
        <main
          class="page__form"
        >
          <form
            class="page__form__container form-container"
          >
            <div
              class="form-container__row row"
            >
              <button
                class="row__image-container"
                data-testid="dropdownButton"
                type="button"
              >
                <img
                  alt="1"
                  class="row__image row__arrow "
                  src="down-arrow.png"
                />
              </button>
              <input
                class="row__textfield"
                placeholder="What needs to be done?"
                type="text"
                value=""
              />
              <button
                class="row__image-container"
                type="submit"
              >
                <img
                  alt="Add task"
                  class="row__image row__image_right-side"
                  src="add-task.svg"
                />
              </button>
            </div>
          </form>
          <ul
            class="page__form-ul "
            data-testid="tasksList"
          />
        </main>
        <div
          class="page__filter filter "
          data-testid="filterBar"
        >
          <span
            class="filter__counter"
          >
             
            items left
          </span>
          <button
            class="filter__button "
            type="button"
          >
            All
          </button>
          <button
            class="filter__button "
            type="button"
          >
            Active
          </button>
          <button
            class="filter__button "
            type="button"
          >
            Completed
          </button>
          <button
            class="filter__button "
            type="button"
          >
            Clear completed
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="page"
    >
      <h1
        class="page__h1"
      >
        todos
      </h1>
      <main
        class="page__form"
      >
        <form
          class="page__form__container form-container"
        >
          <div
            class="form-container__row row"
          >
            <button
              class="row__image-container"
              data-testid="dropdownButton"
              type="button"
            >
              <img
                alt="1"
                class="row__image row__arrow "
                src="down-arrow.png"
              />
            </button>
            <input
              class="row__textfield"
              placeholder="What needs to be done?"
              type="text"
              value=""
            />
            <button
              class="row__image-container"
              type="submit"
            >
              <img
                alt="Add task"
                class="row__image row__image_right-side"
                src="add-task.svg"
              />
            </button>
          </div>
        </form>
        <ul
          class="page__form-ul "
          data-testid="tasksList"
        />
      </main>
      <div
        class="page__filter filter "
        data-testid="filterBar"
      >
        <span
          class="filter__counter"
        >
           
          items left
        </span>
        <button
          class="filter__button "
          type="button"
        >
          All
        </button>
        <button
          class="filter__button "
          type="button"
        >
          Active
        </button>
        <button
          class="filter__button "
          type="button"
        >
          Completed
        </button>
        <button
          class="filter__button "
          type="button"
        >
          Clear completed
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
